In [1]:
import pandas as pd
import matplotlib as plt
import seaborn as sns
from textblob import TextBlob

In [2]:
ls

NLP Lab.ipynb                           q10.png
Womens Clothing E-Commerce Reviews.csv


# Read the data into a dataframe with pandas

In [3]:
clothingpdf=pd.read_csv('Womens Clothing E-Commerce Reviews.csv')

In [4]:
clothingpdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23486 entries, 0 to 23485
Data columns (total 11 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   Unnamed: 0               23486 non-null  int64 
 1   Clothing ID              23486 non-null  int64 
 2   Age                      23486 non-null  int64 
 3   Title                    19676 non-null  object
 4   Review Text              22641 non-null  object
 5   Rating                   23486 non-null  int64 
 6   Recommended IND          23486 non-null  int64 
 7   Positive Feedback Count  23486 non-null  int64 
 8   Division Name            23472 non-null  object
 9   Department Name          23472 non-null  object
 10  Class Name               23472 non-null  object
dtypes: int64(6), object(5)
memory usage: 2.0+ MB


In [5]:
clothingpdf.head()

Unnamed: 0  Clothing ID  Age                    Title  \
0           0          767   33                      NaN   
1           1         1080   34                      NaN   
2           2         1077   60  Some major design flaws   
3           3         1049   50         My favorite buy!   
4           4          847   47         Flattering shirt   

                                         Review Text  Rating  Recommended IND  \
0  Absolutely wonderful - silky and sexy and comf...       4                1   
1  Love this dress!  it's sooo pretty.  i happene...       5                1   
2  I had such high hopes for this dress and reall...       3                0   
3  I love, love, love this jumpsuit. it's fun, fl...       5                1   
4  This shirt is very flattering to all due to th...       5                1   

   Positive Feedback Count   Division Name Department Name Class Name  
0                        0       Initmates        Intimate  Intimates  
1                        4         General         Dresses    Dresses  
2                        0         General         Dresses    Dresses  
3                        0  General Petite         Bottoms      Pants  
4                        6         General            Tops    Blouses

# Filter the data frame to a relevant subset of columns required for our data scenario (image clothing_columns.png)

In [6]:
clothingpdf.drop(['Unnamed: 0', 'Clothing ID', 'Age', 
                                   'Recommended IND', 'Positive Feedback Count',
                                  'Class Name'], axis = 1, inplace = True) 


In [7]:
clothing_columns= clothingpdf

In [8]:
clothing_columns.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23486 entries, 0 to 23485
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Title            19676 non-null  object
 1   Review Text      22641 non-null  object
 2   Rating           23486 non-null  int64 
 3   Division Name    23472 non-null  object
 4   Department Name  23472 non-null  object
dtypes: int64(1), object(4)
memory usage: 917.5+ KB


In [9]:
clothing_columns

Title  \
0                                                    NaN   
1                                                    NaN   
2                                Some major design flaws   
3                                       My favorite buy!   
4                                       Flattering shirt   
...                                                  ...   
23481                     Great dress for many occasions   
23482                         Wish it was made of cotton   
23483                              Cute, but see through   
23484  Very cute dress, perfect for summer parties an...   
23485                    Please make more like this one!   

                                             Review Text  Rating  \
0      Absolutely wonderful - silky and sexy and comf...       4   
1      Love this dress!  it's sooo pretty.  i happene...       5   
2      I had such high hopes for this dress and reall...       3   
3      I love, love, love this jumpsuit. it's fun, fl...       5   
4      This shirt is very flattering to all due to th...       5   
...                                                  ...     ...   
23481  I was very happy to snag this dress at such a ...       5   
23482  It reminds me of maternity clothes. soft, stre...       3   
23483  This fit well, but the top was very see throug...       3   
23484  I bought this dress for a wedding i have this ...       3   
23485  This dress in a lovely platinum is feminine an...       5   

        Division Name Department Name  
0           Initmates        Intimate  
1             General         Dresses  
2             General         Dresses  
3      General Petite         Bottoms  
4             General            Tops  
...               ...             ...  
23481  General Petite         Dresses  
23482  General Petite            Tops  
23483  General Petite         Dresses  
23484         General         Dresses  
23485  General Petite         Dresses  

[23486 rows x 5 columns]

# Use df.columns.str.lower() (or other preferred method) to standardise your column headers


In [10]:
clothing_columns.columns=clothing_columns.columns.str.lower()

# EDA - with descriptive statistics (eg describe(), shape, info()) and/or simple charts, explore and familiarise yourself with the data at your own pace - clarify what each column means/contains and what cleaning steps could be employed (and if needed for our scenario)

In [11]:
clothing_columns.shape

(23486, 5)

In [12]:
clothing_columns.describe()

rating
count  23486.000000
mean       4.196032
std        1.110031
min        1.000000
25%        4.000000
50%        5.000000
75%        5.000000
max        5.000000

In [13]:
clothing_columns.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23486 entries, 0 to 23485
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   title            19676 non-null  object
 1   review text      22641 non-null  object
 2   rating           23486 non-null  int64 
 3   division name    23472 non-null  object
 4   department name  23472 non-null  object
dtypes: int64(1), object(4)
memory usage: 917.5+ KB


In [14]:
clothing_columns

title  \
0                                                    NaN   
1                                                    NaN   
2                                Some major design flaws   
3                                       My favorite buy!   
4                                       Flattering shirt   
...                                                  ...   
23481                     Great dress for many occasions   
23482                         Wish it was made of cotton   
23483                              Cute, but see through   
23484  Very cute dress, perfect for summer parties an...   
23485                    Please make more like this one!   

                                             review text  rating  \
0      Absolutely wonderful - silky and sexy and comf...       4   
1      Love this dress!  it's sooo pretty.  i happene...       5   
2      I had such high hopes for this dress and reall...       3   
3      I love, love, love this jumpsuit. it's fun, fl...       5   
4      This shirt is very flattering to all due to th...       5   
...                                                  ...     ...   
23481  I was very happy to snag this dress at such a ...       5   
23482  It reminds me of maternity clothes. soft, stre...       3   
23483  This fit well, but the top was very see throug...       3   
23484  I bought this dress for a wedding i have this ...       3   
23485  This dress in a lovely platinum is feminine an...       5   

        division name department name  
0           Initmates        Intimate  
1             General         Dresses  
2             General         Dresses  
3      General Petite         Bottoms  
4             General            Tops  
...               ...             ...  
23481  General Petite         Dresses  
23482  General Petite            Tops  
23483  General Petite         Dresses  
23484         General         Dresses  
23485  General Petite         Dresses  

[23486 rows x 5 columns]

# Use the pandas groupby function to summarise the average rating by division and department as a new dataframe (image clothing_rating_groupby.png)

In [15]:
df_by_div=clothing_columns.groupby(['division name', 'department name'],as_index = False).agg({'rating':'mean'})

In [16]:
df_by_div

division name department name    rating
0          General         Bottoms  4.268686
1          General         Dresses  4.163003
2          General         Jackets  4.240310
3          General            Tops  4.148749
4          General           Trend  3.822917
5   General Petite         Bottoms  4.329356
6   General Petite         Dresses  4.133256
7   General Petite        Intimate  4.240343
8   General Petite         Jackets  4.304910
9   General Petite            Tops  4.216469
10  General Petite           Trend  3.782609
11       Initmates        Intimate  4.286285

# This dataframe can be easily visualised as a bar chart - do so now (image clothing_rating_chart.png)

# OPTIONAL - if you find it difficult to do this group by and visualise task in python- you can connect to the csv with Tableau and do the same chart there! this is also a useful exercise to remind you of how to work with Tableau (image clothing_rating_tableau.png)

In [17]:
ls

NLP Lab.ipynb                           q10.png
Womens Clothing E-Commerce Reviews.csv


In [2]:
images = ['q10.png'] 

![alt mtext](q10.png "Title")

# Do a spot /sample check on the review column, index position 5 (hint: iloc/loc/at) to apply textblob over the selected review text (image clothing_sentiment_index5.png) Do this for at least 3 samples to evaluate the accuracy of the sentiment polarity and subjectivity against the text itself

In [20]:
testimonial1= clothing_columns.iloc[5]['review text']

In [21]:
Test_1=testimonial1

In [22]:
testimonial1

'I love tracy reese dresses, but this one is not for the very petite. i am just under 5 feet tall and usually wear a 0p in this brand. this dress was very pretty out of the package but its a lot of dress. the skirt is long and very full so it overwhelmed my small frame. not a stranger to alterations, shortening and narrowing the skirt would take away from the embellishment of the garment. i love the color and the idea of the style but it just did not work on me. i returned this dress.'

In [23]:
Test_1=TextBlob("I love tracy reese dresses, but this one is not for the very petite. i am just under 5 feet tall and usually wear a 0p in this brand. this dress was very pretty out of the package but its a lot of dress. the skirt is long and very full so it overwhelmed my small frame. not a stranger to alterations, shortening and narrowing the skirt would take away from the embellishment of the garment. i love the color and the idea of the style but it just did not work on me. i returned this dress.")



In [24]:
Test_1.sentiment

Sentiment(polarity=0.17874999999999996, subjectivity=0.533125)

In [25]:
testimonial2= clothing_columns.iloc[6]['review text']

In [26]:
Test_2=testimonial2

In [27]:
testimonial2

'I aded this in my basket at hte last mintue to see what it would look like in person. (store pick up). i went with teh darkler color only because i am so pale :-) hte color is really gorgeous, and turns out it mathced everythiing i was trying on with it prefectly. it is a little baggy on me and hte xs is hte msallet size (bummer, no petite). i decided to jkeep it though, because as i said, it matvehd everything. my ejans, pants, and the 3 skirts i waas trying on (of which i ]kept all ) oops.'

In [28]:
Test_2=TextBlob("'I aded this in my basket at hte last mintue to see what it would look like in person. (store pick up). i went with teh darkler color only because i am so pale :-) hte color is really gorgeous, and turns out it mathced everythiing i was trying on with it prefectly. it is a little baggy on me and hte xs is hte msallet size (bummer, no petite). i decided to jkeep it though, because as i said, it matvehd everything. my ejans, pants, and the 3 skirts i waas trying on (of which i ]kept all ) oops.'")


In [29]:
Test_2.sentiment

Sentiment(polarity=0.13375, subjectivity=0.6077777777777778)

In [30]:
testimonial3= clothing_columns.iloc[7]['review text']

In [31]:
Test_3=testimonial3

In [32]:
testimonial3

"I ordered this in carbon for store pick up, and had a ton of stuff (as always) to try on and used this top to pair (skirts and pants). everything went with it. the color is really nice charcoal with shimmer, and went well with pencil skirts, flare pants, etc. my only compaint is it is a bit big, sleeves are long and it doesn't go in petite. also a bit loose for me, but no xxs... so i kept it and wil ldecide later since the light color is already sold out in hte smallest size..."

In [33]:
Test_3=TextBlob("I ordered this in carbon for store pick up, and had a ton of stuff (as always) to try on and used this top to pair (skirts and pants). everything went with it. the color is really nice charcoal with shimmer, and went well with pencil skirts, flare pants, etc. my only compaint is it is a bit big, sleeves are long and it doesn't go in petite. also a bit loose for me, but no xxs... so i kept it and wil ldecide later since the light color is already sold out in hte smallest size...")


In [34]:
Test_3.sentiment

Sentiment(polarity=0.17163461538461539, subjectivity=0.49615384615384617)

# For the any selected customer review, use textblob to break out the text into sentences (image clothing_sentences.png)


In [35]:
clothing_col_textblob = TextBlob("'I love tracy reese dresses, but this one is not for the very petite. i am just under 5 feet tall and usually wear a 0p in this brand. this dress was very pretty out of the package but its a lot of dress. the skirt is long and very full so it overwhelmed my small frame. not a stranger to alterations, shortening and narrowing the skirt would take away from the embellishment of the garment. i love the color and the idea of the style but it just did not work on me. i returned this dress.'")

In [36]:
clothing_col_textblob.sentences

[Sentence("'I love tracy reese dresses, but this one is not for the very petite."),
 Sentence("i am just under 5 feet tall and usually wear a 0p in this brand."),
 Sentence("this dress was very pretty out of the package but its a lot of dress."),
 Sentence("the skirt is long and very full so it overwhelmed my small frame."),
 Sentence("not a stranger to alterations, shortening and narrowing the skirt would take away from the embellishment of the garment."),
 Sentence("i love the color and the idea of the style but it just did not work on me."),
 Sentence("i returned this dress.'")]

In [37]:
clothing_columns.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23486 entries, 0 to 23485
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   title            19676 non-null  object
 1   review text      22641 non-null  object
 2   rating           23486 non-null  int64 
 3   division name    23472 non-null  object
 4   department name  23472 non-null  object
dtypes: int64(1), object(4)
memory usage: 917.5+ KB


# Using dropna, remove any rows in your data which contain have null in the review column ( hint your new data will be 22641 rows)

In [38]:
new_df=clothing_columns.dropna(axis=0, subset=['review text'])

In [39]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 22641 entries, 0 to 23485
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   title            19675 non-null  object
 1   review text      22641 non-null  object
 2   rating           22641 non-null  int64 
 3   division name    22628 non-null  object
 4   department name  22628 non-null  object
dtypes: int64(1), object(4)
memory usage: 1.0+ MB


In [40]:
new_df.reset_index(drop=True, inplace=True)

# Define a function with lambda (or other preferred method) to calculate sentiment polarity for each row of the filtered review data set, as a new column on the data frame. (image clothing_sentiment_allrows.png)

In [45]:
new_df.head()

title                                        review text  \
0                      NaN  Absolutely wonderful - silky and sexy and comf...   
1                      NaN  Love this dress!  it's sooo pretty.  i happene...   
2  Some major design flaws  I had such high hopes for this dress and reall...   
3         My favorite buy!  I love, love, love this jumpsuit. it's fun, fl...   
4         Flattering shirt  This shirt is very flattering to all due to th...   

   rating   division name department name  
0       4       Initmates        Intimate  
1       5         General         Dresses  
2       3         General         Dresses  
3       5  General Petite         Bottoms  
4       5         General            Tops

In [46]:
def sentiment(text):
    try:
        return TextBlob(text).sentiment
    except:
        return None

In [51]:
new_df['polarity']=new_df['review text'].apply(sentiment).apply(lambda x:x[0])

<ipython-input-51-ba75f8e4527c>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['polarity']=new_df['review text'].apply(sentiment).apply(lambda x:x[0])


In [52]:
new_df

title  \
0                                                    NaN   
1                                                    NaN   
2                                Some major design flaws   
3                                       My favorite buy!   
4                                       Flattering shirt   
...                                                  ...   
22636                     Great dress for many occasions   
22637                         Wish it was made of cotton   
22638                              Cute, but see through   
22639  Very cute dress, perfect for summer parties an...   
22640                    Please make more like this one!   

                                             review text  rating  \
0      Absolutely wonderful - silky and sexy and comf...       4   
1      Love this dress!  it's sooo pretty.  i happene...       5   
2      I had such high hopes for this dress and reall...       3   
3      I love, love, love this jumpsuit. it's fun, fl...       5   
4      This shirt is very flattering to all due to th...       5   
...                                                  ...     ...   
22636  I was very happy to snag this dress at such a ...       5   
22637  It reminds me of maternity clothes. soft, stre...       3   
22638  This fit well, but the top was very see throug...       3   
22639  I bought this dress for a wedding i have this ...       3   
22640  This dress in a lovely platinum is feminine an...       5   

        division name department name  polarity  
0           Initmates        Intimate  0.633333  
1             General         Dresses  0.339583  
2             General         Dresses  0.073675  
3      General Petite         Bottoms  0.550000  
4             General            Tops  0.512891  
...               ...             ...       ...  
22636  General Petite         Dresses  0.552667  
22637  General Petite            Tops  0.091667  
22638  General Petite         Dresses  0.414286  
22639         General         Dresses  0.322222  
22640  General Petite         Dresses  0.413889  

[22641 rows x 6 columns]

# using the pandas groupby function again, summarise the minimum review polarity, by division and department as a new dataframe (image clothing_min_review.png) - this means whats the lowest sentiment score seen in each department and division - so everything will be less than 0


In [84]:
new_df_pol = new_df.groupby(['division name','department name'])['polarity'].min().sort_values(ascending=True)
new_df_pol

division name   department name
General         Tops              -0.975000
                Dresses           -0.916667
                Jackets           -0.750000
General Petite  Tops              -0.700000
                Intimate          -0.575000
                Jackets           -0.562500
General         Bottoms           -0.533333
General Petite  Dresses           -0.400000
Initmates       Intimate          -0.392333
General         Trend             -0.270833
General Petite  Bottoms           -0.257778
                Trend             -0.083333
Name: polarity, dtype: float64

# using the pandas groupby function again, summarise the average review polarity, by division and department as a new dataframe


In [68]:
df_by_div1=new_df.groupby(['division name', 'department name'],as_index = False).agg({'polarity':'mean'})

In [69]:
df_by_div1

division name department name  polarity
0          General         Bottoms  0.245849
1          General         Dresses  0.251091
2          General         Jackets  0.237141
3          General            Tops  0.247025
4          General           Trend  0.203986
5   General Petite         Bottoms  0.259258
6   General Petite         Dresses  0.248380
7   General Petite        Intimate  0.239517
8   General Petite         Jackets  0.241773
9   General Petite            Tops  0.256835
10  General Petite           Trend  0.195925
11       Initmates        Intimate  0.256979

# visualise this summary as a simple sorted bar chart

# OPTIONAL - if you find it difficult to do these group by and visualise tasks in python- you can output your data frame to a csv, then connect to that csv with Tableau and do the same charts there! this is also a useful exercise to remind you of how to work with Tableau (image clothing_polarity_tableau.png)

In [71]:
new_df.to_csv('out.csv')


In [76]:
ls

NLP Lab.ipynb                           q10.png
Womens Clothing E-Commerce Reviews.csv  q19.png
out.csv


In [1]:
images1 = ['q19.png'] 


![alt mtext](q19.png "Title")